core/src/avm1/activation.rs

```diff
@@ -0,0 +1,268 @@
+//! Activation records
+
+use crate::avm1::object::Object;
+use crate::avm1::scope::Scope;
+use crate::avm1::Value;
+use crate::tag_utils::SwfSlice;
+use gc_arena::{GcCell, MutationContext};
+use std::cell::{Ref, RefMut};
+use std::sync::Arc;
+
+/// Represents a single activation of a given AVM1 function or keyframe.
+#[derive(Clone)]
+pub struct Activation<'gc> {
+    /// Represents the SWF version of a given function.
+    ///
+    /// Certain AVM1 operations change behavior based on the version of the SWF
+    /// file they were defined in. For example, case sensitivity changes based
+    /// on the SWF version.
+    swf_version: u8,
+
+    /// Action data being executed by the reader below.
+    data: SwfSlice,
+
+    /// The current location of the instruction stream being executed.
+    pc: usize,
+
+    /// All defined local variables in this stack frame.
+    scope: GcCell<'gc, Scope<'gc>>,
+
+    /// The immutable value of `this`.
+    this: GcCell<'gc, Object<'gc>>,
+
+    /// The arguments this function was called by.
+    arguments: Option<GcCell<'gc, Object<'gc>>>,
+
+    /// Indicates if this activation object represents a function or embedded
+    /// block (e.g. ActionWith).
+    is_function: bool,
+
+    /// Local registers, if any.
+    ///
+    /// None indicates a function executing out of the global register set.
+    /// Some indicates the existence of local registers, even if none exist.
+    /// i.e. None(Vec::new()) means no registers should exist at all.
+    ///
+    /// Registers are numbered from 1; r0 does not exist. Therefore this vec,
+    /// while nominally starting from zero, actually starts from r1.
+    ///
+    /// Registers are stored in a `GcCell` so that rescopes (e.g. with) use the
+    /// same register set.
+    local_registers: Option<GcCell<'gc, Vec<Value<'gc>>>>,
+}
+
+unsafe impl<'gc> gc_arena::Collect for Activation<'gc> {
+    #[inline]
+    fn trace(&self, cc: gc_arena::CollectionContext) {
+        self.scope.trace(cc);
+        self.this.trace(cc);
+        self.arguments.trace(cc);
+        self.local_registers.trace(cc);
+    }
+}
+
+impl<'gc> Activation<'gc> {
```

Comment:
Right. Actually, we're even a little further off from ECMAScript spec: Locals are handled by creating a blank object and sticking it on the end of the current scope chain rather than keeping them in the execution context as a separate concern. I don't believe there is any particular observable change from the viewpoint of running code, but I always have to keep that in mind when the spec is talking about scope chains.